src/lib/grammar_ast.rs

```diff
@@ -0,0 +1,171 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt;
+
+pub struct GrammarAST {
```

Comment:
Also, can't be just parse the grammar directly into `Grammar` without having an additional representation `GrammarAST`? Why do we need two?